### PR TITLE
Add C++ example

### PR DESCRIPTION
**Pull request by alexcb**

### CODE REVIEW

File: docs/examples/cpp.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+WORKDIR /code
+
+code:
+  COPY src src
+  SAVE IMAGE
+
+build:
+  FROM +code
+  RUN cmake src
+  # cache cmake temp files to prevent rebuilding .o files when the .cpp files don't change
```
Comment (vladaionescu) [inline docs/examples/cpp.md:80]: Might make sense to split this comment over two lines, so it's easier to read without side-scrolling.

File: docs/examples/cpp.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+We will use CMake to manage the build process of the c++ code, with the following CMakeList.txt file:
+
+```
+cmake_minimum_required(VERSION 2.8.9)
+project (fibonacci)
+add_executable(fibonacci main.cpp fib.cpp)
+```
+
+Cmake caches object files under CmakeFiles which allows cmake to only recompile objects when the corresponding
+source code changes. We will use a mount-based cache to cache these temporary files to allow for faster builds
```
Comment (vladaionescu) [inline docs/examples/cpp.md:57]: Might make sense to link `mount-based cache` to some page that describes a bit more what that is. Either the Earthfile reference, or the Advanced caching techniques page.

File: docs/examples/cpp.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+We will use CMake to manage the build process of the c++ code, with the following CMakeList.txt file:
+
+```
+cmake_minimum_required(VERSION 2.8.9)
+project (fibonacci)
+add_executable(fibonacci main.cpp fib.cpp)
+```
+
+Cmake caches object files under CmakeFiles which allows cmake to only recompile objects when the corresponding
```
Comment (vladaionescu) [inline docs/examples/cpp.md:56]:
```suggestion
CMake caches object files under CmakeFiles which allows cmake to only recompile objects when the corresponding
```

(For consistency)

File: docs/examples/cpp.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+source code changes. We will use a mount-based cache to cache these temporary files to allow for faster builds
+on a local machine. Here's a sample `earth.build` file:
```
Comment (vladaionescu) [inline docs/examples/cpp.md:58]:
```suggestion
on a local machine. Here's a sample `build.earth` file:
```

File: docs/guides/basics.md [hunk 5]
```diff
@@ -128,6 +128,9 @@ public class HelloWorld {
     }
 }
 ```
+```
+```
```
Comment (vladaionescu) [inline docs/guides/basics.md:132]: Not necessary?

File: docs/examples/cpp.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+code:
+  COPY src src
+  SAVE IMAGE
+
+build:
+  FROM +code
+  RUN cmake src
+  # cache cmake temp files to prevent rebuilding .o files when the .cpp files don't change
+  RUN --mount=type=cache,target=/code/CMakeFiles make
+  SAVE ARTIFACT fibonacci AS LOCAL "fibonacci"
```
Comment (vladaionescu) [inline docs/examples/cpp.md:82]:
```suggestion
  SAVE ARTIFACT fibonacci AS LOCAL fibonacci
```

(the quotes are not necessary here)

File: docs/guides/basics.md (hunk 5 above)
Comment (alexcb) [reply docs/guides/basics.md]: whoops, that's sloppy of me. I backed that out.

**Pull request closed without merging**